asterius/src/Asterius/Passes/DataOffsetTable.hs

```diff
@@ -0,0 +1,152 @@
+import qualified Data.Set as Set
+import Data.Tuple
+import Foreign
+import Language.Haskell.GHC.Toolkit.Constants
+
+{-# INLINEABLE sizeofStatic #-}
+sizeofStatic :: AsteriusStatic -> Int
+sizeofStatic = \case
+  SymbolStatic {} -> 8
+  Uninitialized x -> x
+  Serialized buf -> BS.length buf
+
+sizeofStatics :: AsteriusStatics -> Int64
+sizeofStatics =
+  fromIntegral
+    . getSum
+    . foldMap (Sum . sizeofStatic)
+    . asteriusStatics
+
+{-# INLINEABLE makeDataOffsetTable #-}
+makeDataOffsetTable :: AsteriusModule -> (SM.SymbolMap Int64, Int64)
+makeDataOffsetTable AsteriusModule {..} =
+  swap $
+    SM.mapAccum
+      ( \a ss -> (a + fromIntegral (fromIntegral (sizeofStatics ss) `roundup` 16), a)
+      )
+      0
+      staticsMap
+
+makeWasmApplyRelocs :: Set.Set Int32 -> Set.Set Int32 -> AsteriusModule
+makeWasmApplyRelocs fn_offsets ss_offsets = runEDSL "__wasm_apply_relocs" $ do
+  setReturnTypes [] -- Zero outputs
+  _ <- params [] -- Zero inputs
+  table_base <-
+    i64Local
+      $ extendUInt32
+      $ GetGlobal
+        { globalSymbol = "__asterius_table_base",
+          valueType = I32
+        }
+  memory_base <-
+    i64Local
+      $ extendUInt32
+      $ GetGlobal
+        { globalSymbol = "__asterius_memory_base",
+          valueType = I32
+        }
+  for_ fn_offsets $ \off ->
+    let loc = memory_base
+                `addInt64` constI64 (fromIntegral off)
+                `addInt64` constI64 (fromIntegral $ dataTag `shiftL` 32) -- TODO: compute this more intuitively
+
+     in storeI64 loc 0 $ table_base `addInt64` loadI64 loc 0 `addInt64` constI64 (fromIntegral $ functionTag `shiftL` 32)
+  for_ ss_offsets $ \off ->
+    let loc = memory_base
+                `addInt64` constI64 (fromIntegral off)
+                `addInt64` constI64 (fromIntegral $ dataTag `shiftL` 32) -- TODO: compute this more intuitively
+     in storeI64 loc 0 $ memory_base `addInt64` loadI64 loc 0 `addInt64` constI64 (fromIntegral $ dataTag `shiftL` 32)
+
+-- | Given the offset of a static and the static itself, compute the
+-- corresponding data segment and the offset of the subsequent static.
+-- Furthermore, gather the offsets of symbol statics, to be used by the
+-- relocation function. NOTE: we do not generate data segments for
+-- uninitialized statics; we do not have to specify each segment and the
+-- linear memory is zero-initialized anyway.
+{-# INLINEABLE makeSegment #-}
+makeSegment ::
+  SM.SymbolMap Int64 ->
+  SM.SymbolMap Int64 ->
+  (Int32, Set.Set Int32, Set.Set Int32) ->
+  AsteriusStatic ->
+  ((Int32, Set.Set Int32, Set.Set Int32), Bag DataSegment)
+makeSegment fn_off_map ss_off_map (current_off, fn_meta, ss_meta) static = case static of
+  SymbolStatic sym o
+    | Just off <- SM.lookup sym fn_off_map ->
+      let off_to_store = off + fromIntegral o
+       in ( (next_off, current_off `Set.insert` fn_meta, ss_meta),
+            unitBag DataSegment {content = encodeStorable off_to_store, offset = current_off}
```

Comment:
Looks like another critical error to me. In the `DataSegment`s we produce in this module, the `offset`s start from `0` right? (Based on `makeDataOffsetTable`). Well, they really should be `memory_base + offset`. And this also implies that the `DataSegment` should make the `offset` field an `Expression` type.

Reply:
Ah, indeed. I'll do the generalization in a separate PR, like for the `FunctionTable`.

Reply:
#777 is merged, would you solve the conflict here and see if the corrected offsets solve the crashes?

Reply:
Will do.